tsdb/chunks/head_chunks_test.go

```diff
@@ -0,0 +1,301 @@
+	expectedData := []expectedDataType{}
+	numChunks := 100
+
+	timesRan := 0
+	populateChunks := func() {
+		for i := 0; i < numChunks; i++ {
+			chunk := randomChunk(t)
+			chunkRef, err := hrw.WriteChunk(
+				uint64(rand.Int63()),
+				int64(((timesRan*numChunks)+i)*1000+1),
+				int64(((timesRan*numChunks)+i+1)*1000),
+				chunk,
+			)
+			testutil.Ok(t, err)
+
+			expectedData = append(expectedData, expectedDataType{
+				chunkRef: chunkRef,
+				chunk:    chunk,
+			})
+		}
+	}
+
+	// Add chunks till we fill at least 1 segment file
+	// and also have something in the in-memory buffer.
+	for hrw.curFileSequence < 2 || hrw.wbuf.Buffered() == 0 {
+		populateChunks()
+		timesRan++
+	}
+
+	// This is to test the access of buffered chunks.
+	testutil.Assert(t, hrw.wbuf.Buffered() > 0, "there are no buffered chunks")
+
+	for _, exp := range expectedData {
+		actChunk, err := hrw.Chunk(exp.chunkRef)
+		testutil.Ok(t, err)
+		testutil.Equals(t, exp.chunk.Bytes(), actChunk.Bytes())
+	}
+}
+
+func TestHeadReadWriter_IterateChunks(t *testing.T) {
+	hrw, close := testHeadReadWriter(t)
+	defer func(h *ChunkDiskMapper) {
+		testutil.Ok(t, h.Close())
+		close()
+	}(hrw)
+
+	type expectedDataType struct {
+		seriesRef, chunkRef uint64
+		mint, maxt          int64
+		chunk               chunkenc.Chunk
+	}
+
+	expectedData := []expectedDataType{}
+	numChunks := 10000
+
+	for i := 0; i < numChunks; i++ {
+		seriesRef := uint64(rand.Int63())
+		mint := int64(i*1000 + 1)
+		maxt := int64((i + 1) * 1000)
+		chunk := randomChunk(t)
+
+		chunkRef, err := hrw.WriteChunk(seriesRef, mint, maxt, chunk)
+		testutil.Ok(t, err)
+
+		expectedData = append(expectedData, expectedDataType{
+			seriesRef: seriesRef,
+			mint:      mint,
+			maxt:      maxt,
+			chunkRef:  chunkRef,
+			chunk:     chunk,
+		})
+	}
+
+	dir := hrw.dir.Name()
+	testutil.Ok(t, hrw.Close())
+
+	hrw, err := NewChunkDiskMapper(dir, chunkenc.NewPool())
+	testutil.Ok(t, err)
+	defer func() {
+		testutil.Ok(t, hrw.Close())
+	}()
+
+	expIdx := 0
+	err = hrw.IterateAllChunks(func(seriesRef, chunkRef uint64, mint, maxt int64) error {
+		t.Helper()
+
+		expData := expectedData[expIdx]
+		testutil.Equals(t, expData.seriesRef, seriesRef)
+		testutil.Equals(t, expData.chunkRef, chunkRef)
+		testutil.Equals(t, expData.maxt, maxt)
+		testutil.Equals(t, expData.maxt, maxt)
+
+		actChunk, err := hrw.Chunk(expData.chunkRef)
+		testutil.Ok(t, err)
+		testutil.Equals(t, expData.chunk.Bytes(), actChunk.Bytes())
+
+		expIdx++
+		return nil
+	})
+	testutil.Ok(t, err)
+	testutil.Equals(t, expIdx, len(expectedData))
+}
+
+func TestHeadReadWriter_Truncate(t *testing.T) {
+	hrw, close := testHeadReadWriter(t)
+	defer func() {
+		testutil.Ok(t, hrw.Close())
+		close()
+	}()
+
+	var timeToTruncate int64
+
+	// Cut 5 segments.
+	for i := 1; i <= 5; i++ {
+		testutil.Ok(t, hrw.cut(int64(i-1)*100))
+
+		// Write a chunks to set maxt for the segment.
+		_, err := hrw.WriteChunk(1, int64((i-1)*100)+1, int64(i*100)-1, randomChunk(t))
```

Comment:
hm , are you sure this can't be simplified a bit? maybe it is just me, but I don't want to do match when reading tests :)

Reply:
It can be simplified by not using `i` and just incrementing some variable :P will make the change

Reply:
Using the `i` felt intuitive to be TBH at you can guess the ranges of the previous and the next file

Reply:
much better, thanks.